[TOPIC-GPIO] update sam0 driver

pabigot opened the pull request:
Tested on SAM-D21-Xplained Pro with 2-pin test, blinky, and button.

Checked D20 and R21 data sheets, there should be no incompatibilities.

pabigot committed:
drivers: gpio_sam0: update to new GPIO API

Drop the port access, rework to separate interrupt and pin
configuration, add new API.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
boards: atmel sam0: update devicetree to new GPIO flags

Use standard spelling for pull and active flags.  Correct LED active
level (based on SAM-D21-Xplained-Pro).

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
@mnkp This does not pass the (updated) 1-pin test for interrupts.  I believe that's because on SAM0 a pin that is configured for interrupts set for an alternate pinmux function (EIC) and cannot be set with the GPIO interface.

Although I've confirmed that setting and clearing the pin results in a change in the input value before `gpio_pin_interrupt_configure()`, while the interrupts are configured no changes to the input value can be detected through the GPIO interface, and only one interrupt is ever signaled in the loop that toggles the GPIO.

The two pin interrupt tests pass.

jfischer-no reviewed (approved):
Tested on samr21xpro with `tests/drivers/gpio/gpio_basic_api`, blinky and button samples. :heavy_check_mark:  LGTM

benpicco commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -230,10 +270,14 @@ static const struct gpio_driver_api gpio_sam0_api = {
+	.port_set_bits_raw = gpio_sam0_port_set_bits_raw,
+	.port_clear_bits_raw = gpio_sam0_port_clear_bits_raw,
+	.port_toggle_bits = gpio_sam0_port_toggle_bits,
+	.pin_interrupt_configure = gpio_sam0_pin_interrupt_configure,
```
This should be behind `#ifdef CONFIG_SAM0_EIC`

pabigot commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -47,122 +49,65 @@ static int gpio_sam0_config(struct device *dev, int access_op, u32_t pin,
-				trigger = SAM0_EIC_RISING;
+		/* Pull configuration is supported if not output */
+		if ((flags & (GPIO_PULL_UP | GPIO_PULL_DOWN)) != 0) {
+			pincfg.bit.PULLEN = 1;
```
That's how it works: `pincfg` is being built up, and is only applied at the end.

pabigot commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -172,52 +117,147 @@ static int gpio_sam0_read(struct device *dev, int access_op, u32_t pin,
+	};
+	enum sam0_eic_trigger trigger;
+	int rc = 0;
+
```
I think we only need to check `INEN`; interrupts should still work in bidirectional mode.

mnkp replied:
The documentation could be more clear about it, however we do need to configure pin as function A in pinmux. This disconnects the PORT module. So I doubt the bidirectional mode would work. Also your earlier comments confirms that

> Although I've confirmed that setting and clearing the pin results in a change in the input value before gpio_pin_interrupt_configure(), while the interrupts are configured no changes to the input value can be detected through the GPIO interface

pabigot replied:
Correct, you can't do bidirectional mode when the interrupts are active, but you can have the GPIO configured that way underneath so that when interrupts are turned off the pin returns to bidirectional mode.  So `INEN` is enough to gate entry to interrupt mode.  I've confirmed this by running the tests with the input pin configured bidirectional.

mnkp replied:
> you can have the GPIO configured that way underneath so that when interrupts are turned off the pin returns to bidirectional mode.

We shouldn't do it. The input shouldn't suddenly become an output unless explicitly configured via `gpio_pin_configure` function. That's dangerous.

pabigot replied:
I don't understand.   The way this API model works is (a) configure the GPIO (direction, pull, debounce), then (b) configure (and enable) interrupts.

Deconfiguring (disabling) interrupts should restore the underlying GPIO configuration, whether it was input-only, output only (where that works), or bidirectional.  Do you think it should do something else?

mnkp replied:
the (b) cannot change output mode to input and vice versa. The `gpio_pin_interrupt_configure` function should only configure interrupts.

pabigot replied:
Um, yeah, it is.  DIRSET and INEN, works fine.

mnkp replied:
Does the driver pass 1 pin testcase?

pabigot replied:
> Does the driver pass 1 pin testcase?

Everything but the interrupt tests.

The 2-pin test passes when the input GPIO is configured as disconnected.  So on this device the GPIO is completely disabled (except possibly for the effect of pull settings) when interrupts are enabled.  The check I added for `INEN` being set is wrong: you don't need the input buffer enabled support external interrupts.

So I'm not sure what to do.  The 1-pin test for interrupts simply can't work on this device, but that can't be detected without improperly rejecting GPIO/interrupt configurations that are supported.  Perhaps the 1-pin test should skip the interrupt section if `IS_ENABLED(CONFIG_GPIO_SAM0)`.

mnkp replied:
> The 2-pin test passes when the input GPIO is configured as disconnected.

OK, that was not obvious. According to the documentation `INEN` has to be enabled for the input signal to be routed to other modules. But documentation is not always perfect.

> So I'm not sure what to do. The 1-pin test for interrupts simply can't work on this device, but that can't be detected without improperly rejecting GPIO/interrupt configurations that are supported. 

If the 1-pin testcase doesn't work then clearly some GPIO/interrupt configurations are not supported. As I stated already `gpio_pin_interrupt_configure` cannot change pin mode from input to output or vice versa. This is reserved for `gpio_pin_configure` function. If GPIO pin is configured as an output and then `gpio_pin_interrupt_configure` connects it to EIC module by means of pinmux function A the pin becomes an input.

pabigot replied:
> If the 1-pin testcase doesn't work then clearly some GPIO/interrupt configurations are not supported. As I stated already `gpio_pin_interrupt_configure` cannot change pin mode from input to output or vice versa. This is reserved for `gpio_pin_configure` function. If GPIO pin is configured as an output and then `gpio_pin_interrupt_configure` connects it to EIC module by means of pinmux function A the pin becomes an input.

I don't think that model is correct for this driver. `gpio_pin_interrupt_configure()` will transform a disconnected GPIO to an "input" even if `INEN` is not set.

However, I'm done arguing for consistency.  I've pushed a commit that performs the validation you want, and now the 1-pin interrupt test is skipped on this platform.

mnkp replied:
> `gpio_pin_interrupt_configure()` will transform a disconnected GPIO to an "input" even if `INEN` is not set.

This isn't allowed. The `gpio_pin_interrupt_configure()` can't change pin drive settings. It should only verify that the pin drive settings are compatible with the interrupt mode, i.e.  that the pin is an input. This is the case now, finally, so that's good.

We need to change however the comment you've added to the code as it is not correct and confusing. I propose to write something like:
"Enabling pin interrupt requires disconnecting it from the I/O pin controller (PORT) module and connecting to the External Interrupt Controller (EIC). As a result interrupts are only supported if the pin is configured in input mode."

pabigot replied:
Is the new wording acceptable?

mnkp replied:
Looks good!

pabigot commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -47,122 +49,65 @@ static int gpio_sam0_config(struct device *dev, int access_op, u32_t pin,
-		regs->DIRSET.bit.DIRSET = mask;
+	PORT_PINCFG_Type pincfg = {
+		.reg = 0,
+	};
```
Assigning to `.reg` makes it comparable with the parallel statement in interrupt configuration where we're modifying, not replacing, the configuration.

pabigot commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -172,52 +117,147 @@ static int gpio_sam0_read(struct device *dev, int access_op, u32_t pin,
+	enum sam0_eic_trigger trigger;
+	int rc = 0;
+
+	switch (mode) {
```
I don't understand how that would work.  The disable path needs to reverse all the changes made in the enable path; that's the only thing that will restore GPIO operation.  Changes can be made when we have pinmuxing code to add.

mnkp commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -172,22 +117,151 @@ static int gpio_sam0_read(struct device *dev, int access_op, u32_t pin,
+	case GPIO_INT_MODE_EDGE:
+		/* Interrupt detection requires input buffer */
+		if (pincfg.bit.INEN == 0) {
+			rc = -EINVAL;
```
We should return `-ENOTSUP` since the configuration is valid, it's however not supported by the hardware. That's also what other drivers are doing. With `-EINVAL` the 1 pin testcase will not pass.

mnkp commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -47,122 +49,65 @@ static int gpio_sam0_config(struct device *dev, int access_op, u32_t pin,
+	/* Supports disconnected, input, output, or bidirectional */
+	if ((flags & GPIO_INPUT) != 0) {
+		pincfg.bit.INEN = 1;
 	}
```
we need `} else {` here. Or we should use `WRITE_BIT` macro.

pabigot replied:
`pincfg` is zeroed above, so we only need to set this, not clear it.

mnkp replied:
Sorry, I keep forgetting we're not writing to a register but a variable.

benpicco commented on drivers/gpio/gpio_sam0.c:
```diff
@@ -172,22 +117,158 @@ static int gpio_sam0_read(struct device *dev, int access_op, u32_t pin,
+		/* Transfer control to EIC */
+		pincfg.bit.PMUXEN = 1;
+		if ((pin & 1U) != 0) {
+			regs->PMUX[pin / 2U].bit.PMUXO = PORT_PMUX_PMUXE_A_Val;
```
If you change this to `PORT_PMUX_PMUXE(0)` it will just work™ on same54 too.

`samd21/component/port.h`:
```C
#define   PORT_PMUX_PMUXE_A_Val           _U_(0x0)
#define PORT_PMUX_PMUXE_A           (PORT_PMUX_PMUXE_A_Val)
#define PORT_PMUX_PMUXE(value)      (PORT_PMUX_PMUXE_Msk & ((value) << PORT_PMUX_PMUXE_Pos))
```

`same54/component/port.h` only contains the last define.

benpicco replied:
ac7010d375fca893d2975119786f9a6b1d7d0985 should work too.

pabigot replied:
Thanks.  I've cherry-picked that commit into this branch so we should be good no matter which PR gets in first.

benpicco committed:
gpio: sam0: provide PORT_PMUX_PMUXE_A_Val define

SAMD5x/SAME5x header files do not provide this define anymore.
On SAMD2x it was 0, this is still valid.

Signed-off-by: Benjamin Valentin <benpicco@googlemail.com>
